content/advent-2015/glow-map-reduce-for-golang.md

```diff
@@ -0,0 +1,485 @@
+```
+
+I hope you like the code here. It may not look as simple as other
+languages that are skipping data types. But when a project gets
+reasonably large, readability is a big deal. Glow's anonymous
+functions have the right amount of type information, which helps to
+understand the code, especially when someone else wrote it.
+
+If you do not have much data now, you can actually stop here and run
+Glow in standalone mode. Glow's API works for both standalone
+mode and distributed mode. I highly encourage you start using Glow
+in standalone mode. It should just work.
+
+## Go channels in Glow
+Glow works natually with channels. The data flows from one dataset
+to the next dataset via channels, either in standalone mode or
+distributed mode.
+
+Another interesting usage of channel is that a channel can be a
+mapper's result emitter. Usually 1 mapper emits 1 result, which
+can be represented as function's return result, e.g.,
+``` func(key string) int ```. But if a mapper can emit 0 or more
+than 1 result, this won't work. The channel fits nicely here. If the
+last parameter of a mapper funtion is a channel, e.g.,
+``` func(line string, ch chan string) ```, it will be treated
+as the result emitter.
+
+# Setup Glow Cluster
+
+Now let us setup the Glow cluster, just in case you need to scale up.
+Setting up Glow Cluster is super easy. First, build the "glow" binary file:
+
+```
+$ go get github.com/chrislusf/glow
+```
+
+The compiled binary file is usually ```$GOPATH/bin/glow```.
+Now copy it to any computer, and run this:
+```
+// start a master on one computer
+> glow master
+// run one or more agents on computers
+> glow agent --dir ./glow/data --max.executors=16 --memory=2048 --master="localhost:8930" --port 8931
+// it's fine to run several agents on the same computer
+> glow agent --dir ./glow/data2 --max.executors=8 --memory=512 --master="localhost:8930" --port 8932
+```
+
+Either master or agent only takes 5~6 MB memory. They are quiet
+and efficient. I highly recommend run this "glow agent" on any
+machine you can find, so that you can tap into the compute power any
+time you want, with one line of code change as follows.
+
+# Distributed Execution
+
+Just insert this line to the import section, around line 7:
+```go
+    _ "github.com/chrislusf/glow/driver"
+```
+
+
+Now the word count can run distributedly this way:
+```
+$ go run t.go -glow
+2015/12/08 00:54:46 localhost:8930 allocated 1 executors.
+2015/12/08 00:54:46 localhost:8930 allocated 1 executors.
+2015/12/08 00:54:46 localhost:8930 allocated 2 executors.
+2015/12/08 00:54:46 localhost:8930 allocated 1 executors.
+127.0.0.1:8931>count: 532
+```
+
+
+Also, you can visualize the flow. Run this command:
+```
+$ go run t.go -glow -glow.flow.plot > x.dot
+$ dot -Tpng -owc.png x.dot
+```
+
+And the flow graph looks like this:
+
+![Word Count Flow](/postimages/advent-2015/glow-word-count-flow.png)
+
+## Something wrong!
+
+You may get a different result when running distributedly vs
```

Comment:
I think you have to go with either:
"distributedly vs in standalone"
or
"in distributed vs standalone"